# Taylor Summations of sin(x)

In [2]:
import math
import time
from decimal import Decimal

def taylorSin(x,n):
    start_time = time.time()
    s = 0
    for k in range(0,n+1):
        s += (Decimal(((-1)**k))*(Decimal((x**((2*k)+1)))/(Decimal(math.factorial((2*k)+1)))))
    end_time = time.time()
    total_time = end_time - start_time
    s = float(s)
    return s, total_time

def main():
    X=[0.03125,0.0625,0.125,0.25,0.5,1,2,4]
    N=[1,2,3,4,8,16,32,64,128,256]
    taylorSins=[[0.0 for i in range(len(X))] for j in range(len(N))]
    taylorSinTimes=[[0.0 for i in range(len(X))] for j in range(len(N))]
    sinExactValues=[math.sin(x) for x in X]

    for i in range(len(N)):
        for j in range(len(X)):
            taylorsin, taylorsintime = taylorSin(X[j],N[i])
            taylorSins[i][j] = taylorsin
            taylorSinTimes[i][j] = taylorsintime
        
    for i in range(len(X)):
        for j in range(len(N)):
            if (X[i] <= 0.0625):
                print(f'x={X[i]}\tn={N[j]}\t     sin(x)={taylorSins[j][i]}\t\ttime={taylorSinTimes[j][i]}')
            else:
                print(f'x={X[i]}\t\tn={N[j]}\t     sin(x)={taylorSins[j][i]}\t\ttime={taylorSinTimes[j][i]}')
        if (X[i] <= 0.0625):
            print(f'x={X[i]}\t\tExact Value={sinExactValues[i]}\n')
        else:
            print(f'x={X[i]}\t\t\tExact Value={sinExactValues[i]}\n')

main()

x=0.03125	n=1	     sin(x)=0.031244913736979168		time=0.0005617141723632812
x=0.03125	n=2	     sin(x)=0.031244913985331853		time=0.0007278919219970703
x=0.03125	n=3	     sin(x)=0.03124491398532608		time=0.0007259845733642578
x=0.03125	n=4	     sin(x)=0.03124491398532608		time=0.0009000301361083984
x=0.03125	n=8	     sin(x)=0.03124491398532608		time=0.0020487308502197266
x=0.03125	n=16	     sin(x)=0.03124491398532608		time=0.006940603256225586
x=0.03125	n=32	     sin(x)=0.03124491398532608		time=0.009717226028442383
x=0.03125	n=64	     sin(x)=0.03124491398532608		time=0.008149862289428711
x=0.03125	n=128	     sin(x)=0.03124491398532608		time=0.007417440414428711
x=0.03125	n=256	     sin(x)=0.03124491398532608		time=0.011333703994750977
x=0.03125		Exact Value=0.03124491398532608

x=0.0625	n=1	     sin(x)=0.062459309895833336		time=0.001917123794555664
x=0.0625	n=2	     sin(x)=0.062459317843119304		time=0.0005373954772949219
x=0.0625	n=3	     sin(x)=0.06245931784238016		time=0.000771284103

# Taylor Summations of sin(x)

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.031244913736979168</td>
    <td class="tg-baqh">0.062459309895833336</td>
    <td class="tg-baqh">0.12467447916666667</td>
    <td class="tg-baqh">0.24739583333333334</td>
    <td class="tg-baqh">0.4791666666666667</td>
    <td class="tg-baqh">0.8333333333333334</td>
    <td class="tg-baqh">0.6666666666666666</td>
    <td class="tg-baqh">-6.666666666666667</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.031244913985331853</td>
    <td class="tg-baqh">0.062459317843119304</td>
    <td class="tg-baqh">0.12467473347981771</td>
    <td class="tg-baqh">0.24740397135416667</td>
    <td class="tg-baqh">0.47942708333333334</td>
    <td class="tg-baqh">0.8416666666666667</td>
    <td class="tg-baqh">0.9333333333333333</td>
    <td class="tg-baqh">1.8666666666666667</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.06245931784238016</td>
    <td class="tg-baqh">0.12467473338520717</td>
    <td class="tg-baqh">0.24740395924401662</td>
    <td class="tg-baqh">0.479425533234127</td>
    <td class="tg-baqh">0.841468253968254</td>
    <td class="tg-baqh">0.9079365079365079</td>
    <td class="tg-baqh">-1.384126984126984</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.2474039592545289</td>
    <td class="tg-baqh">0.4794255386164159</td>
    <td class="tg-baqh">0.8414710097001764</td>
    <td class="tg-baqh">0.909347442680776</td>
    <td class="tg-baqh">-0.6617283950617284</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.24740395925452294</td>
    <td class="tg-baqh">0.479425538604203</td>
    <td class="tg-baqh">0.8414709848078965</td>
    <td class="tg-baqh">0.909297426829951</td>
    <td class="tg-baqh">-0.7568003190686041</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.24740395925452294</td>
    <td class="tg-baqh">0.479425538604203</td>
    <td class="tg-baqh">0.8414709848078965</td>
    <td class="tg-baqh">0.9092974268256817</td>
    <td class="tg-baqh">-0.7568024953079282</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.24740395925452294</td>
    <td class="tg-baqh">0.479425538604203</td>
    <td class="tg-baqh">0.8414709848078965</td>
    <td class="tg-baqh">0.9092974268256817</td>
    <td class="tg-baqh">-0.7568024953079282</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.24740395925452294</td>
    <td class="tg-baqh">0.479425538604203</td>
    <td class="tg-baqh">0.8414709848078965</td>
    <td class="tg-baqh">0.9092974268256817</td>
    <td class="tg-baqh">-0.7568024953079282</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.24740395925452294</td>
    <td class="tg-baqh">0.479425538604203</td>
    <td class="tg-baqh">0.8414709848078965</td>
    <td class="tg-baqh">0.9092974268256817</td>
    <td class="tg-baqh">-0.7568024953079282</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.24740395925452294</td>
    <td class="tg-baqh">0.479425538604203</td>
    <td class="tg-baqh">0.8414709848078965</td>
    <td class="tg-baqh">0.9092974268256817</td>
    <td class="tg-baqh">-0.7568024953079282</td>
  </tr>
  <tr>
    <td class="tg-baqh">Exact Value</td>
    <td class="tg-baqh">0.03124491398532608</td>
    <td class="tg-baqh">0.0624593178423802</td>
    <td class="tg-baqh">0.12467473338522769</td>
    <td class="tg-baqh">0.24740395925452294</td>
    <td class="tg-baqh">0.479425538604203</td>
    <td class="tg-baqh">0.8414709848078965</td>
    <td class="tg-baqh">0.9092974268256817</td>
    <td class="tg-baqh">-0.7568024953079282</td>
  </tr>
</table>

# Taylor Summation of sin(x) Calculation Times

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.0005097389221191406</td>
    <td class="tg-baqh">0.00020766258239746094</td>
    <td class="tg-baqh">0.00019598007202148438</td>
    <td class="tg-baqh">0.00019121170043945312</td>
    <td class="tg-baqh">0.0001876354217529297</td>
    <td class="tg-baqh">0.00015306472778320312</td>
    <td class="tg-baqh">0.0001499652862548828</td>
    <td class="tg-baqh">0.00014901161193847656</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.00029850006103515625</td>
    <td class="tg-baqh">0.00028586387634277344</td>
    <td class="tg-baqh">0.0002853870391845703</td>
    <td class="tg-baqh">0.0002834796905517578</td>
    <td class="tg-baqh">0.00028443336486816406</td>
    <td class="tg-baqh">0.00029015541076660156</td>
    <td class="tg-baqh">0.00022864341735839844</td>
    <td class="tg-baqh">0.0002269744873046875</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.0003936290740966797</td>
    <td class="tg-baqh">0.0003829002380371094</td>
    <td class="tg-baqh">0.0003809928894042969</td>
    <td class="tg-baqh">0.0003757476806640625</td>
    <td class="tg-baqh">0.00037932395935058594</td>
    <td class="tg-baqh">0.0003159046173095703</td>
    <td class="tg-baqh">0.0003199577331542969</td>
    <td class="tg-baqh">0.000301361083984375</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.000492095947265625</td>
    <td class="tg-baqh">0.0004782676696777344</td>
    <td class="tg-baqh">0.00047659873962402344</td>
    <td class="tg-baqh">0.00047659873962402344</td>
    <td class="tg-baqh">0.0004277229309082031</td>
    <td class="tg-baqh">0.0002987384796142578</td>
    <td class="tg-baqh">0.0003147125244140625</td>
    <td class="tg-baqh">0.0002849102020263672</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0006666183471679688</td>
    <td class="tg-baqh">0.0006566047668457031</td>
    <td class="tg-baqh">0.0006515979766845703</td>
    <td class="tg-baqh">0.0006437301635742188</td>
    <td class="tg-baqh">0.0006630420684814453</td>
    <td class="tg-baqh">0.0005500316619873047</td>
    <td class="tg-baqh">0.0005435943603515625</td>
    <td class="tg-baqh">0.0005266666412353516</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0013020038604736328</td>
    <td class="tg-baqh">0.0013210773468017578</td>
    <td class="tg-baqh">0.001062631607055664</td>
    <td class="tg-baqh">0.0010094642639160156</td>
    <td class="tg-baqh">0.0009975433349609375</td>
    <td class="tg-baqh">0.0008478164672851562</td>
    <td class="tg-baqh">0.0008471012115478516</td>
    <td class="tg-baqh">0.0008437633514404297</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0020394325256347656</td>
    <td class="tg-baqh">0.0019600391387939453</td>
    <td class="tg-baqh">0.0016427040100097656</td>
    <td class="tg-baqh">0.0016548633575439453</td>
    <td class="tg-baqh">0.0016224384307861328</td>
    <td class="tg-baqh">0.0014011859893798828</td>
    <td class="tg-baqh">0.0013947486877441406</td>
    <td class="tg-baqh">0.0013339519500732422</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0031011104583740234</td>
    <td class="tg-baqh">0.0030450820922851562</td>
    <td class="tg-baqh">0.002912282943725586</td>
    <td class="tg-baqh">0.0026721954345703125</td>
    <td class="tg-baqh">0.0026578903198242188</td>
    <td class="tg-baqh">0.0023794174194335938</td>
    <td class="tg-baqh">0.00220489501953125</td>
    <td class="tg-baqh">0.0022242069244384766</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.005771636962890625</td>
    <td class="tg-baqh">0.006316661834716797</td>
    <td class="tg-baqh">0.0058841705322265625</td>
    <td class="tg-baqh">0.005759239196777344</td>
    <td class="tg-baqh">0.005769968032836914</td>
    <td class="tg-baqh">0.004782915115356445</td>
    <td class="tg-baqh">0.004868268966674805</td>
    <td class="tg-baqh">0.0049152374267578125</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.010770320892333984</td>
    <td class="tg-baqh">0.012658357620239258</td>
    <td class="tg-baqh">0.012682914733886719</td>
    <td class="tg-baqh">0.015440940856933594</td>
    <td class="tg-baqh">0.014740467071533203</td>
    <td class="tg-baqh">0.012476682662963867</td>
    <td class="tg-baqh">0.012889623641967773</td>
    <td class="tg-baqh">0.01308298110961914</td>
  </tr>
</table>

# Taylor Summation of sin(x) Absolute Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">2.483469124792137e-10</td>
    <td class="tg-baqh">7.946546864978643e-09</td>
    <td class="tg-baqh">2.542185610215908e-07</td>
    <td class="tg-baqh">8.125921189594543e-06</td>
    <td class="tg-baqh">0.0002588719375363202</td>
    <td class="tg-baqh">0.008137651474563135</td>
    <td class="tg-baqh">0.24263076015901508</td>
    <td class="tg-baqh">5.9098641713587385</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">-5.773159728050814e-15</td>
    <td class="tg-baqh">-7.391032230685823e-13</td>
    <td class="tg-baqh">-9.459001637512898e-11</td>
    <td class="tg-baqh">-1.2099643731389165e-08</td>
    <td class="tg-baqh">-1.5447291303316568e-06</td>
    <td class="tg-baqh">-0.0001956818587701692</td>
    <td class="tg-baqh">-0.02403590650765164</td>
    <td class="tg-baqh">-2.6234691619745947</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">4.163336342344337e-17</td>
    <td class="tg-baqh">2.052524816775758e-14</td>
    <td class="tg-baqh">1.0506318037784013e-11</td>
    <td class="tg-baqh">5.370075994992618e-09</td>
    <td class="tg-baqh">2.7308396425285153e-06</td>
    <td class="tg-baqh">0.001360918889173779</td>
    <td class="tg-baqh">0.6273244888190559</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">-5.9674487573602164e-15</td>
    <td class="tg-baqh">-1.2212897360086572e-11</td>
    <td class="tg-baqh">-2.4892279903987458e-08</td>
    <td class="tg-baqh">-5.001585509434303e-05</td>
    <td class="tg-baqh">-0.09507410024619978</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">-4.269251618893577e-12</td>
    <td class="tg-baqh">-2.176239324080953e-06</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Taylor Summation of sin(x) Relative Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">7.94839488339789e-07</td>
    <td class="tg-baqh">1.2722758029557989e-05</td>
    <td class="tg-baqh">0.00020390585364447178</td>
    <td class="tg-baqh">0.003284582880846636</td>
    <td class="tg-baqh">0.05402544783366682</td>
    <td class="tg-baqh">0.976518176947576</td>
    <td class="tg-baqh">36.39461402385226</td>
    <td class="tg-baqh">88.64796257038108</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">1.8477118326397263e-11</td>
    <td class="tg-baqh">1.1833354070965154e-09</td>
    <td class="tg-baqh">7.586943539802407e-08</td>
    <td class="tg-baqh">4.89064248450084e-06</td>
    <td class="tg-baqh">0.00032220314288286593</td>
    <td class="tg-baqh">0.023249329754871588</td>
    <td class="tg-baqh">2.5752756972483897</td>
    <td class="tg-baqh">140.54299082006756</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">6.665676933665472e-14</td>
    <td class="tg-baqh">1.646303754614079e-11</td>
    <td class="tg-baqh">4.246624859960928e-09</td>
    <td class="tg-baqh">1.1201063820624978e-06</td>
    <td class="tg-baqh">0.00032453269979589055</td>
    <td class="tg-baqh">0.14989141611529383</td>
    <td class="tg-baqh">45.322755499541884</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">2.4120263779695264e-12</td>
    <td class="tg-baqh">2.547402333912379e-09</td>
    <td class="tg-baqh">2.9581862734471148e-06</td>
    <td class="tg-baqh">0.0055001919779853565</td>
    <td class="tg-baqh">14.367541268548848</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">4.695110194886734e-10</td>
    <td class="tg-baqh">0.00028755792898703527</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Taylor Summation of cos(x)

In [1]:
import math
import time
from decimal import Decimal

def taylorCos(x,n):
    start_time = time.time()
    s = 0
    for k in range(0,n+1):
        s += (Decimal(((-1)**k))*(Decimal((x**(2*k)))/(Decimal(math.factorial(2*k)))))
    end_time = time.time()
    total_time = end_time - start_time
    s = float(s)
    return s, total_time

def main():
    X=[0.03125,0.0625,0.125,0.25,0.5,1,2,4]
    N=[1,2,3,4,8,16,32,64,128,256]
    taylorCoss=[[0.0 for i in range(len(X))] for j in range(len(N))]
    taylorCosTimes=[[0.0 for i in range(len(X))] for j in range(len(N))]
    cosExactValues=[math.cos(x) for x in X]

    for i in range(len(N)):
        for j in range(len(X)):
            taylorcos, taylorcostime = taylorCos(X[j],N[i])
            taylorCoss[i][j] = taylorcos
            taylorCosTimes[i][j] = taylorcostime
    
    for i in range(len(X)):
        for j in range(len(N)):
            if (X[i] <= 0.0625):
                print(f'x={X[i]}\tn={N[j]}\t     cos(x)={taylorCoss[j][i]}\t\ttime={taylorCosTimes[j][i]}')
            else:
                print(f'x={X[i]}\t\tn={N[j]}\t     cos(x)={taylorCoss[j][i]}\t\ttime={taylorCosTimes[j][i]}')
        if (X[i] <= 0.0625):
            print(f'x={X[i]}\t\tExact Value={cosExactValues[i]}\n')
        else:
            print(f'x={X[i]}\t\t\tExact Value={cosExactValues[i]}\n')

main()

x=0.03125	n=1	     cos(x)=0.99951171875		time=0.0007109642028808594
x=0.03125	n=2	     cos(x)=0.9995117584864298		time=0.0004436969757080078
x=0.03125	n=3	     cos(x)=0.9995117584851363		time=0.0005588531494140625
x=0.03125	n=4	     cos(x)=0.9995117584851364		time=0.000782012939453125
x=0.03125	n=8	     cos(x)=0.9995117584851364		time=0.001302480697631836
x=0.03125	n=16	     cos(x)=0.9995117584851364		time=0.002896547317504883
x=0.03125	n=32	     cos(x)=0.9995117584851364		time=0.0053768157958984375
x=0.03125	n=64	     cos(x)=0.9995117584851364		time=0.006997585296630859
x=0.03125	n=128	     cos(x)=0.9995117584851364		time=0.0067098140716552734
x=0.03125	n=256	     cos(x)=0.9995117584851364		time=0.010439872741699219
x=0.03125		Exact Value=0.9995117584851364

x=0.0625	n=1	     cos(x)=0.998046875		time=0.0005064010620117188
x=0.0625	n=2	     cos(x)=0.9980475107828776		time=0.0004150867462158203
x=0.0625	n=3	     cos(x)=0.9980475107000933		time=0.0007243156433105469
x=0.0625	n=4	     cos

# Taylor Summation of cos(x)

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.99951171875</td>
    <td class="tg-baqh">0.998046875</td>
    <td class="tg-baqh">0.9921875</td>
    <td class="tg-baqh">0.96875</td>
    <td class="tg-baqh">0.875</td>
    <td class="tg-baqh">0.5</td>
    <td class="tg-baqh">-1.0</td>
    <td class="tg-baqh">-7.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.9995117584864298</td>
    <td class="tg-baqh">0.9980475107828776</td>
    <td class="tg-baqh">0.9921976725260416</td>
    <td class="tg-baqh">0.9689127604166666</td>
    <td class="tg-baqh">0.8776041666666666</td>
    <td class="tg-baqh">0.5416666666666666</td>
    <td class="tg-baqh">-0.3333333333333333</td>
    <td class="tg-baqh">3.6666666666666665</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.9995117584851363</td>
    <td class="tg-baqh">0.9980475107000933</td>
    <td class="tg-baqh">0.992197667227851</td>
    <td class="tg-baqh">0.9689124213324652</td>
    <td class="tg-baqh">0.8775824652777777</td>
    <td class="tg-baqh">0.5402777777777777</td>
    <td class="tg-baqh">-0.4222222222222222</td>
    <td class="tg-baqh">-2.022222222222222</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.9921976672293293</td>
    <td class="tg-baqh">0.9689124217109074</td>
    <td class="tg-baqh">0.8775825621589782</td>
    <td class="tg-baqh">0.5403025793650794</td>
    <td class="tg-baqh">-0.4158730158730159</td>
    <td class="tg-baqh">-0.3968253968253968</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.992197667229329</td>
    <td class="tg-baqh">0.9689124217106447</td>
    <td class="tg-baqh">0.8775825618903728</td>
    <td class="tg-baqh">0.5403023058681399</td>
    <td class="tg-baqh">-0.4161468365066249</td>
    <td class="tg-baqh">-0.6536333241518427</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.992197667229329</td>
    <td class="tg-baqh">0.9689124217106447</td>
    <td class="tg-baqh">0.8775825618903728</td>
    <td class="tg-baqh">0.5403023058681398</td>
    <td class="tg-baqh">-0.4161468365471424</td>
    <td class="tg-baqh">-0.6536436208636119</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.992197667229329</td>
    <td class="tg-baqh">0.9689124217106447</td>
    <td class="tg-baqh">0.8775825618903728</td>
    <td class="tg-baqh">0.5403023058681398</td>
    <td class="tg-baqh">-0.4161468365471424</td>
    <td class="tg-baqh">-0.6536436208636119</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.992197667229329</td>
    <td class="tg-baqh">0.9689124217106447</td>
    <td class="tg-baqh">0.8775825618903728</td>
    <td class="tg-baqh">0.5403023058681398</td>
    <td class="tg-baqh">-0.4161468365471424</td>
    <td class="tg-baqh">-0.6536436208636119</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.992197667229329</td>
    <td class="tg-baqh">0.9689124217106447</td>
    <td class="tg-baqh">0.8775825618903728</td>
    <td class="tg-baqh">0.5403023058681398</td>
    <td class="tg-baqh">-0.4161468365471424</td>
    <td class="tg-baqh">-0.6536436208636119</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.992197667229329</td>
    <td class="tg-baqh">0.9689124217106447</td>
    <td class="tg-baqh">0.8775825618903728</td>
    <td class="tg-baqh">0.5403023058681398</td>
    <td class="tg-baqh">-0.4161468365471424</td>
    <td class="tg-baqh">-0.6536436208636119</td>
  </tr>
  <tr>
    <td class="tg-baqh">Exact Value</td>
    <td class="tg-baqh">0.9995117584851364</td>
    <td class="tg-baqh">0.9980475107000991</td>
    <td class="tg-baqh">0.992197667229329</td>
    <td class="tg-baqh">0.9689124217106447</td>
    <td class="tg-baqh">0.8775825618903728</td>
    <td class="tg-baqh">0.5403023058681398</td>
    <td class="tg-baqh">-0.4161468365471424</td>
    <td class="tg-baqh">-0.6536436208636119</td>
  </tr>
</table>

# Taylor Summation of cos(x) Calculation Times

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.0007064342498779297</td>
    <td class="tg-baqh">0.00038504600524902344</td>
    <td class="tg-baqh">0.00035643577575683594</td>
    <td class="tg-baqh">0.00035309791564941406</td>
    <td class="tg-baqh">0.0003514289855957031</td>
    <td class="tg-baqh">0.00029087066650390625</td>
    <td class="tg-baqh">0.00028634071350097656</td>
    <td class="tg-baqh">0.0003116130828857422</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.0006051063537597656</td>
    <td class="tg-baqh">0.0005674362182617188</td>
    <td class="tg-baqh">0.0005555152893066406</td>
    <td class="tg-baqh">0.0005545616149902344</td>
    <td class="tg-baqh">0.0005524158477783203</td>
    <td class="tg-baqh">0.0005519390106201172</td>
    <td class="tg-baqh">0.00045108795166015625</td>
    <td class="tg-baqh">0.0004429817199707031</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.0007793903350830078</td>
    <td class="tg-baqh">0.0007617473602294922</td>
    <td class="tg-baqh">0.000782012939453125</td>
    <td class="tg-baqh">0.0007522106170654297</td>
    <td class="tg-baqh">0.0007457733154296875</td>
    <td class="tg-baqh">0.0006346702575683594</td>
    <td class="tg-baqh">0.0006163120269775391</td>
    <td class="tg-baqh">0.0006182193756103516</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.0009758472442626953</td>
    <td class="tg-baqh">0.0008132457733154297</td>
    <td class="tg-baqh">0.0008080005645751953</td>
    <td class="tg-baqh">0.0008683204650878906</td>
    <td class="tg-baqh">0.0008058547973632812</td>
    <td class="tg-baqh">0.0006773471832275391</td>
    <td class="tg-baqh">0.0006687641143798828</td>
    <td class="tg-baqh">0.0006589889526367188</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0014994144439697266</td>
    <td class="tg-baqh">0.0015022754669189453</td>
    <td class="tg-baqh">0.0014941692352294922</td>
    <td class="tg-baqh">0.0014781951904296875</td>
    <td class="tg-baqh">0.0014798641204833984</td>
    <td class="tg-baqh">0.0012485980987548828</td>
    <td class="tg-baqh">0.0012454986572265625</td>
    <td class="tg-baqh">0.0013594627380371094</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0027008056640625</td>
    <td class="tg-baqh">0.002480030059814453</td>
    <td class="tg-baqh">0.0024933815002441406</td>
    <td class="tg-baqh">0.002495288848876953</td>
    <td class="tg-baqh">0.0027244091033935547</td>
    <td class="tg-baqh">0.00237274169921875</td>
    <td class="tg-baqh">0.0021371841430664062</td>
    <td class="tg-baqh">0.002120494842529297</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.005126237869262695</td>
    <td class="tg-baqh">0.005293369293212891</td>
    <td class="tg-baqh">0.005044221878051758</td>
    <td class="tg-baqh">0.0049037933349609375</td>
    <td class="tg-baqh">0.004958391189575195</td>
    <td class="tg-baqh">0.004286766052246094</td>
    <td class="tg-baqh">0.004179239273071289</td>
    <td class="tg-baqh">0.004163503646850586</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.01075291633605957</td>
    <td class="tg-baqh">0.009732961654663086</td>
    <td class="tg-baqh">0.0066754817962646484</td>
    <td class="tg-baqh">0.005192279815673828</td>
    <td class="tg-baqh">0.005536794662475586</td>
    <td class="tg-baqh">0.0034210681915283203</td>
    <td class="tg-baqh">0.003389120101928711</td>
    <td class="tg-baqh">0.003056049346923828</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.007472038269042969</td>
    <td class="tg-baqh">0.006854057312011719</td>
    <td class="tg-baqh">0.00597834587097168</td>
    <td class="tg-baqh">0.005608081817626953</td>
    <td class="tg-baqh">0.005505800247192383</td>
    <td class="tg-baqh">0.004645347595214844</td>
    <td class="tg-baqh">0.004750490188598633</td>
    <td class="tg-baqh">0.004824638366699219</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.010611534118652344</td>
    <td class="tg-baqh">0.01140904426574707</td>
    <td class="tg-baqh">0.01243448257446289</td>
    <td class="tg-baqh">0.015202045440673828</td>
    <td class="tg-baqh">0.01432037353515625</td>
    <td class="tg-baqh">0.012246131896972656</td>
    <td class="tg-baqh">0.012468576431274414</td>
    <td class="tg-baqh">0.012806892395019531</td>
  </tr>
</table>

# Taylor Summation of cos(x) Absolute Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">3.9735136403429294e-08</td>
    <td class="tg-baqh">6.357000991163986e-07</td>
    <td class="tg-baqh">1.01672293290056e-05</td>
    <td class="tg-baqh">0.00016242171064473343</td>
    <td class="tg-baqh">0.0025825618903727587</td>
    <td class="tg-baqh">0.040302305868139765</td>
    <td class="tg-baqh">0.5838531634528576</td>
    <td class="tg-baqh">6.346356379136388</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">-1.2934098236883074e-12</td>
    <td class="tg-baqh">-8.27784507606566e-11</td>
    <td class="tg-baqh">-5.296712624058841e-09</td>
    <td class="tg-baqh">-3.38706021896229e-07</td>
    <td class="tg-baqh">-2.160477629387092e-05</td>
    <td class="tg-baqh">-0.0013643607985268646</td>
    <td class="tg-baqh">-0.08281350321380909</td>
    <td class="tg-baqh">-4.320310287530279</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">1.1102230246251565e-16</td>
    <td class="tg-baqh">5.773159728050814e-15</td>
    <td class="tg-baqh">1.478039912683471e-12</td>
    <td class="tg-baqh">3.7817948772556065e-10</td>
    <td class="tg-baqh">9.661259503523922e-08</td>
    <td class="tg-baqh">2.4528090362019306e-05</td>
    <td class="tg-baqh">0.006075385675079814</td>
    <td class="tg-baqh">1.3685786013586103</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">-3.3306690738754696e-16</td>
    <td class="tg-baqh">-2.6267876762631204e-13</td>
    <td class="tg-baqh">-2.686054711986685e-10</td>
    <td class="tg-baqh">-2.734969396511744e-07</td>
    <td class="tg-baqh">-0.00027382067412651256</td>
    <td class="tg-baqh">-0.25681822403821514</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">-1.1102230246251565e-16</td>
    <td class="tg-baqh">-4.0517533772543857e-11</td>
    <td class="tg-baqh">-1.0296711769219868e-05</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Taylor Summation of cos(x) Relative Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">3.9754547803724075e-06</td>
    <td class="tg-baqh">6.369441306215188e-05</td>
    <td class="tg-baqh">0.0010247286252856038</td>
    <td class="tg-baqh">0.016766112066553127</td>
    <td class="tg-baqh">0.29514993032831527</td>
    <td class="tg-baqh">8.060461173627953</td>
    <td class="tg-baqh">58.385316345285766</td>
    <td class="tg-baqh">90.66223398766269</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">1.2940416285316445e-10</td>
    <td class="tg-baqh">8.294039097970839e-09</td>
    <td class="tg-baqh">5.338364290428046e-07</td>
    <td class="tg-baqh">3.495732905308972e-05</td>
    <td class="tg-baqh">0.0024617905331888526</td>
    <td class="tg-baqh">0.2518819935741904</td>
    <td class="tg-baqh">24.844050964142728</td>
    <td class="tg-baqh">117.82664420537124</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">1.1107653463805314e-14</td>
    <td class="tg-baqh">5.784453812224988e-13</td>
    <td class="tg-baqh">1.4896627572336852e-10</td>
    <td class="tg-baqh">3.903133858119825e-08</td>
    <td class="tg-baqh">1.100894774654127e-05</td>
    <td class="tg-baqh">0.004539903614563985</td>
    <td class="tg-baqh">1.4389071335715349</td>
    <td class="tg-baqh">67.67696380344776</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">3.356860415905053e-14</td>
    <td class="tg-baqh">2.7110682218571762e-11</td>
    <td class="tg-baqh">3.0607430318335035e-08</td>
    <td class="tg-baqh">5.0619217841337394e-05</td>
    <td class="tg-baqh">0.06584237583958126</td>
    <td class="tg-baqh">64.71819245763022</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">2.0548182241074962e-14</td>
    <td class="tg-baqh">9.736355107892029e-09</td>
    <td class="tg-baqh">0.0015753039798851937</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Taylor Summation of e<sup>x</sup>

In [1]:
import math
import time
from decimal import Decimal

def taylorExp(x,n):
    start_time = time.time()
    s = 0
    for k in range(0,n+1):
        s += (Decimal(x**k)/Decimal(math.factorial(k)))
    end_time = time.time()
    total_time = end_time - start_time
    s = float(s)
    return s, total_time

def main():
    X=[0.03125,0.0625,0.125,0.25,0.5,1,2,4]
    N=[1,2,3,4,8,16,32,64,128,256]
    taylorExps=[[0.0 for i in range(len(X))] for j in range(len(N))]
    taylorExpTimes=[[0.0 for i in range(len(X))] for j in range(len(N))]
    expExactValues=[math.exp(x) for x in X]

    for i in range(len(N)):
        for j in range(len(X)):
            taylorexp, taylorexptime = taylorExp(X[j],N[i])
            taylorExps[i][j] = taylorexp
            taylorExpTimes[i][j] = taylorexptime
    
    for i in range(len(X)):
        for j in range(len(N)):
            if (X[i] <= 0.0625):
                print(f'x={X[i]}\tn={N[j]}\t     exp(x)={taylorExps[j][i]}\t\ttime={taylorExpTimes[j][i]}')
            else:
                print(f'x={X[i]}\t\tn={N[j]}\t     exp(x)={taylorExps[j][i]}\t\ttime={taylorExpTimes[j][i]}')
        if (X[i] <= 0.0625):
            print(f'x={X[i]}\t\tExact Value={expExactValues[i]}\n')
        else:
            print(f'x={X[i]}\t\t\tExact Value={expExactValues[i]}\n')

main()

x=0.03125	n=1	     exp(x)=1.03125		time=0.0004620552062988281
x=0.03125	n=2	     exp(x)=1.03173828125		time=0.0007276535034179688
x=0.03125	n=3	     exp(x)=1.0317433675130208		time=0.0005664825439453125
x=0.03125	n=4	     exp(x)=1.0317434072494507		time=0.0008459091186523438
x=0.03125	n=8	     exp(x)=1.0317434074991028		time=0.0010809898376464844
x=0.03125	n=16	     exp(x)=1.0317434074991028		time=0.002101898193359375
x=0.03125	n=32	     exp(x)=1.0317434074991028		time=0.0045926570892333984
x=0.03125	n=64	     exp(x)=1.0317434074991028		time=0.009319543838500977
x=0.03125	n=128	     exp(x)=1.0317434074991028		time=0.006325483322143555
x=0.03125	n=256	     exp(x)=1.0317434074991028		time=0.011158943176269531
x=0.03125		Exact Value=1.0317434074991028

x=0.0625	n=1	     exp(x)=1.0625		time=0.0004329681396484375
x=0.0625	n=2	     exp(x)=1.064453125		time=0.007008075714111328
x=0.0625	n=3	     exp(x)=1.0644938151041667		time=0.0006124973297119141
x=0.0625	n=4	     exp(x)=1.0644944508870442	

# Taylor Summation of e<sup>x</sup>

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">1.03125</td>
    <td class="tg-baqh">1.0625</td>
    <td class="tg-baqh">1.125</td>
    <td class="tg-baqh">1.25</td>
    <td class="tg-baqh">1.5</td>
    <td class="tg-baqh">2.0</td>
    <td class="tg-baqh">3.0</td>
    <td class="tg-baqh">5.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">1.03173828125</td>
    <td class="tg-baqh">1.064453125</td>
    <td class="tg-baqh">1.1328125</td>
    <td class="tg-baqh">1.28125</td>
    <td class="tg-baqh">1.625</td>
    <td class="tg-baqh">2.5</td>
    <td class="tg-baqh">5.0</td>
    <td class="tg-baqh">13.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">1.0317433675130208</td>
    <td class="tg-baqh">1.0644938151041667</td>
    <td class="tg-baqh">1.1331380208333333</td>
    <td class="tg-baqh">1.2838541666666667</td>
    <td class="tg-baqh">1.6458333333333333</td>
    <td class="tg-baqh">2.6666666666666665</td>
    <td class="tg-baqh">6.333333333333333</td>
    <td class="tg-baqh">23.666666666666668</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">1.0317434072494507</td>
    <td class="tg-baqh">1.0644944508870442</td>
    <td class="tg-baqh">1.133148193359375</td>
    <td class="tg-baqh">1.2840169270833333</td>
    <td class="tg-baqh">1.6484375</td>
    <td class="tg-baqh">2.7083333333333335</td>
    <td class="tg-baqh">7.0</td>
    <td class="tg-baqh">34.333333333333336</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">1.0317434074991028</td>
    <td class="tg-baqh">1.0644944589178593</td>
    <td class="tg-baqh">1.1331484530668055</td>
    <td class="tg-baqh">1.2840254166769602</td>
    <td class="tg-baqh">1.6487212650359624</td>
    <td class="tg-baqh">2.71827876984127</td>
    <td class="tg-baqh">7.387301587301588</td>
    <td class="tg-baqh">53.43174603174603</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">1.0317434074991028</td>
    <td class="tg-baqh">1.0644944589178593</td>
    <td class="tg-baqh">1.1331484530668263</td>
    <td class="tg-baqh">1.2840254166877414</td>
    <td class="tg-baqh">1.6487212707001282</td>
    <td class="tg-baqh">2.7182818284590424</td>
    <td class="tg-baqh">7.3890560985164155</td>
    <td class="tg-baqh">54.59808818263845</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">1.0317434074991028</td>
    <td class="tg-baqh">1.0644944589178593</td>
    <td class="tg-baqh">1.1331484530668263</td>
    <td class="tg-baqh">1.2840254166877414</td>
    <td class="tg-baqh">1.6487212707001282</td>
    <td class="tg-baqh">2.718281828459045</td>
    <td class="tg-baqh">7.38905609893065</td>
    <td class="tg-baqh">54.598150033144236</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">1.0317434074991028</td>
    <td class="tg-baqh">1.0644944589178593</td>
    <td class="tg-baqh">1.1331484530668263</td>
    <td class="tg-baqh">1.2840254166877414</td>
    <td class="tg-baqh">1.6487212707001282</td>
    <td class="tg-baqh">2.718281828459045</td>
    <td class="tg-baqh">7.38905609893065</td>
    <td class="tg-baqh">54.598150033144236</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">1.0317434074991028</td>
    <td class="tg-baqh">1.0644944589178593</td>
    <td class="tg-baqh">1.1331484530668263</td>
    <td class="tg-baqh">1.2840254166877414</td>
    <td class="tg-baqh">1.6487212707001282</td>
    <td class="tg-baqh">2.718281828459045</td>
    <td class="tg-baqh">7.38905609893065</td>
    <td class="tg-baqh">54.598150033144236</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">1.0317434074991028</td>
    <td class="tg-baqh">1.0644944589178593</td>
    <td class="tg-baqh">1.1331484530668263</td>
    <td class="tg-baqh">1.2840254166877414</td>
    <td class="tg-baqh">1.6487212707001282</td>
    <td class="tg-baqh">2.718281828459045</td>
    <td class="tg-baqh">7.38905609893065</td>
    <td class="tg-baqh">54.598150033144236</td>
  </tr>
  <tr>
    <td class="tg-baqh">Exact Value</td>
    <td class="tg-baqh">1.0317434074991028</td>
    <td class="tg-baqh">1.0644944589178593</td>
    <td class="tg-baqh">1.1331484530668263</td>
    <td class="tg-baqh">1.2840254166877414</td>
    <td class="tg-baqh">1.6487212707001282</td>
    <td class="tg-baqh">2.718281828459045</td>
    <td class="tg-baqh">7.38905609893065</td>
    <td class="tg-baqh">54.598150033144236</td>
  </tr>
</table>

# Taylor Summation of e<sup>x</sup> Calculation Times

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.0004608631134033203</td>
    <td class="tg-baqh">0.0002703666687011719</td>
    <td class="tg-baqh">0.00022482872009277344</td>
    <td class="tg-baqh">0.00022172927856445312</td>
    <td class="tg-baqh">0.000217437744140625</td>
    <td class="tg-baqh">0.00017523765563964844</td>
    <td class="tg-baqh">0.0001728534698486328</td>
    <td class="tg-baqh">0.00017023086547851562</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.0003361701965332031</td>
    <td class="tg-baqh">0.0003311634063720703</td>
    <td class="tg-baqh">0.0003299713134765625</td>
    <td class="tg-baqh">0.00032901763916015625</td>
    <td class="tg-baqh">0.0003256797790527344</td>
    <td class="tg-baqh">0.0003371238708496094</td>
    <td class="tg-baqh">0.0003554821014404297</td>
    <td class="tg-baqh">0.0003407001495361328</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.0005877017974853516</td>
    <td class="tg-baqh">0.0005636215209960938</td>
    <td class="tg-baqh">0.0004913806915283203</td>
    <td class="tg-baqh">0.0005002021789550781</td>
    <td class="tg-baqh">0.0004961490631103516</td>
    <td class="tg-baqh">0.00038623809814453125</td>
    <td class="tg-baqh">0.00035881996154785156</td>
    <td class="tg-baqh">0.0003581047058105469</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.0005552768707275391</td>
    <td class="tg-baqh">0.000392913818359375</td>
    <td class="tg-baqh">0.00038814544677734375</td>
    <td class="tg-baqh">0.0003914833068847656</td>
    <td class="tg-baqh">0.00039958953857421875</td>
    <td class="tg-baqh">0.0003235340118408203</td>
    <td class="tg-baqh">0.00030803680419921875</td>
    <td class="tg-baqh">0.00029730796813964844</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0007202625274658203</td>
    <td class="tg-baqh">0.0007116794586181641</td>
    <td class="tg-baqh">0.0007536411285400391</td>
    <td class="tg-baqh">0.0008270740509033203</td>
    <td class="tg-baqh">0.0007126331329345703</td>
    <td class="tg-baqh">0.0005953311920166016</td>
    <td class="tg-baqh">0.0005788803100585938</td>
    <td class="tg-baqh">0.0005900859832763672</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0012307167053222656</td>
    <td class="tg-baqh">0.0010268688201904297</td>
    <td class="tg-baqh">0.0010368824005126953</td>
    <td class="tg-baqh">0.0010154247283935547</td>
    <td class="tg-baqh">0.0010089874267578125</td>
    <td class="tg-baqh">0.0008387565612792969</td>
    <td class="tg-baqh">0.0008554458618164062</td>
    <td class="tg-baqh">0.0008196830749511719</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.002007722854614258</td>
    <td class="tg-baqh">0.0016155242919921875</td>
    <td class="tg-baqh">0.0016179084777832031</td>
    <td class="tg-baqh">0.0015888214111328125</td>
    <td class="tg-baqh">0.001596212387084961</td>
    <td class="tg-baqh">0.0013422966003417969</td>
    <td class="tg-baqh">0.00128173828125</td>
    <td class="tg-baqh">0.0010905265808105469</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0027282238006591797</td>
    <td class="tg-baqh">0.0027031898498535156</td>
    <td class="tg-baqh">0.0026254653930664062</td>
    <td class="tg-baqh">0.0022656917572021484</td>
    <td class="tg-baqh">0.0022640228271484375</td>
    <td class="tg-baqh">0.0018877983093261719</td>
    <td class="tg-baqh">0.001875162124633789</td>
    <td class="tg-baqh">0.0017402172088623047</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.004594326019287109</td>
    <td class="tg-baqh">0.004481315612792969</td>
    <td class="tg-baqh">0.004302024841308594</td>
    <td class="tg-baqh">0.004248380661010742</td>
    <td class="tg-baqh">0.004245758056640625</td>
    <td class="tg-baqh">0.0038776397705078125</td>
    <td class="tg-baqh">0.004979372024536133</td>
    <td class="tg-baqh">0.003833293914794922</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.01198267936706543</td>
    <td class="tg-baqh">0.012152671813964844</td>
    <td class="tg-baqh">0.011117935180664062</td>
    <td class="tg-baqh">0.010102987289428711</td>
    <td class="tg-baqh">0.009421586990356445</td>
    <td class="tg-baqh">0.007698535919189453</td>
    <td class="tg-baqh">0.007863521575927734</td>
    <td class="tg-baqh">0.007970094680786133</td>
  </tr>
</table>

# Taylor Summation of e<sup>x</sup> Absolute Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.0004934074991027604</td>
    <td class="tg-baqh">0.001994458917859321</td>
    <td class="tg-baqh">0.008148453066826322</td>
    <td class="tg-baqh">0.034025416687741394</td>
    <td class="tg-baqh">0.1487212707001282</td>
    <td class="tg-baqh">0.7182818284590451</td>
    <td class="tg-baqh">4.38905609893065</td>
    <td class="tg-baqh">49.598150033144236</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">5.126249102760383e-06</td>
    <td class="tg-baqh">4.1333917859320835e-05</td>
    <td class="tg-baqh">0.0003359530668263222</td>
    <td class="tg-baqh">0.0027754166877413944</td>
    <td class="tg-baqh">0.023721270700128194</td>
    <td class="tg-baqh">0.2182818284590451</td>
    <td class="tg-baqh">2.3890560989306504</td>
    <td class="tg-baqh">41.598150033144236</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">3.998608200106446e-08</td>
    <td class="tg-baqh">6.43813692580153e-07</td>
    <td class="tg-baqh">1.0432233493062881e-05</td>
    <td class="tg-baqh">0.0001712500210746537</td>
    <td class="tg-baqh">0.002887937366794935</td>
    <td class="tg-baqh">0.05161516179237857</td>
    <td class="tg-baqh">1.0557227655973174</td>
    <td class="tg-baqh">30.93148336647757</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">2.496520767891752e-10</td>
    <td class="tg-baqh">8.030815124016044e-09</td>
    <td class="tg-baqh">2.5970745132219974e-07</td>
    <td class="tg-baqh">8.489604408135065e-06</td>
    <td class="tg-baqh">0.00028377070012819416</td>
    <td class="tg-baqh">0.00994849512571161</td>
    <td class="tg-baqh">0.3890560989306504</td>
    <td class="tg-baqh">20.2648166998109</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">2.0872192862952943e-14</td>
    <td class="tg-baqh">1.078115374752997e-11</td>
    <td class="tg-baqh">5.6641658030542885e-09</td>
    <td class="tg-baqh">3.0586177750535626e-06</td>
    <td class="tg-baqh">0.001754511629062705</td>
    <td class="tg-baqh">1.166404001398206</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">2.6645352591003757e-15</td>
    <td class="tg-baqh">4.142348686286823e-10</td>
    <td class="tg-baqh">6.185050578721984e-05</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Taylor Summation of e<sup>x</sup> Relative Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
    <th class="tg-baqh">x=1</th>
    <th class="tg-baqh">x=2</th>
    <th class="tg-baqh">x=4</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.047845575670570704</td>
    <td class="tg-baqh">0.18771378050440665</td>
    <td class="tg-baqh">0.7243069392734508</td>
    <td class="tg-baqh">2.7220333350193116</td>
    <td class="tg-baqh">9.914751380008546</td>
    <td class="tg-baqh">35.914091422952254</td>
    <td class="tg-baqh">146.301869964355</td>
    <td class="tg-baqh">991.9630006628846</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.0004968555685022841</td>
    <td class="tg-baqh">0.0038831130172426176</td>
    <td class="tg-baqh">0.029656546588806374</td>
    <td class="tg-baqh">0.2166178878237186</td>
    <td class="tg-baqh">1.4597705046232736</td>
    <td class="tg-baqh">8.731273138361804</td>
    <td class="tg-baqh">47.78112197861301</td>
    <td class="tg-baqh">319.98576948572486</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">3.875584109394319e-06</td>
    <td class="tg-baqh">6.048073586196949e-05</td>
    <td class="tg-baqh">0.000920649850350163</td>
    <td class="tg-baqh">0.013338744035023737</td>
    <td class="tg-baqh">0.17546961215969226</td>
    <td class="tg-baqh">1.9355685672141965</td>
    <td class="tg-baqh">16.6693068252208</td>
    <td class="tg-baqh">130.69640859075028</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">2.419710899386589e-08</td>
    <td class="tg-baqh">7.544252689456444e-07</td>
    <td class="tg-baqh">2.2919107389851716e-05</td>
    <td class="tg-baqh">0.0006611754276027614</td>
    <td class="tg-baqh">0.01721452588455396</td>
    <td class="tg-baqh">0.3673290507955056</td>
    <td class="tg-baqh">5.557944270437863</td>
    <td class="tg-baqh">59.02373796061426</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">1.841964555170461e-12</td>
    <td class="tg-baqh">8.396370981060052e-10</td>
    <td class="tg-baqh">3.4354902330508484e-07</td>
    <td class="tg-baqh">0.00011252038639260557</td>
    <td class="tg-baqh">0.02375037228855832</td>
    <td class="tg-baqh">2.1829793858976583</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">9.802277420994514e-14</td>
    <td class="tg-baqh">5.606059327548655e-09</td>
    <td class="tg-baqh">0.00011328328123930094</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Taylor Summation of log<sub>e</sub>(1-x)

In [1]:
import math
import time
from decimal import Decimal

def taylorLog(x,n):
    start_time = time.time()
    s = 0
    for k in range(1,n+1):
        s -= (Decimal(x**k)/Decimal(k))
    end_time = time.time()
    total_time = end_time - start_time
    s = float(s)
    return s, total_time

def main():
    X=[0.03125,0.0625,0.125,0.25,0.5]
    N=[1,2,3,4,8,16,32,64,128,256]
    taylorLogs=[[0.0 for i in range(len(X))] for j in range(len(N))]
    taylorLogTimes=[[0.0 for i in range(len(X))] for j in range(len(N))]
    logExactValues=[math.log(1-x) for x in X]

    for i in range(len(N)):
        for j in range(len(X)):
            taylorlog, taylorlogtime = taylorLog(X[j],N[i])
            taylorLogs[i][j] = taylorlog
            taylorLogTimes[i][j] = taylorlogtime
    
    for i in range(len(X)):
        for j in range(len(N)):
            if (X[i] <= 0.0625):
                print(f'x={X[i]}\tn={N[j]}\t     log(x)={taylorLogs[j][i]}\t\ttime={taylorLogTimes[j][i]}')
            else:
                print(f'x={X[i]}\t\tn={N[j]}\t     log(x)={taylorLogs[j][i]}\t\ttime={taylorLogTimes[j][i]}')
        if (X[i] <= 0.0625):
            print(f'x={X[i]}\t\tExact Value={logExactValues[i]}\n')
        else:
            print(f'x={X[i]}\t\t\tExact Value={logExactValues[i]}\n')

main()

x=0.03125	n=1	     log(x)=-0.03125		time=0.00042724609375
x=0.03125	n=2	     log(x)=-0.03173828125		time=0.00054168701171875
x=0.03125	n=3	     log(x)=-0.031748453776041664		time=0.001018524169921875
x=0.03125	n=4	     log(x)=-0.031748692194620766		time=0.0018019676208496094
x=0.03125	n=8	     log(x)=-0.03174869831457705		time=0.002323627471923828
x=0.03125	n=16	     log(x)=-0.0317486983145803		time=0.00400090217590332
x=0.03125	n=32	     log(x)=-0.0317486983145803		time=0.004594087600708008
x=0.03125	n=64	     log(x)=-0.0317486983145803		time=0.01873922348022461
x=0.03125	n=128	     log(x)=-0.0317486983145803		time=0.009417533874511719
x=0.03125	n=256	     log(x)=-0.0317486983145803		time=0.018643617630004883
x=0.03125		Exact Value=-0.0317486983145803

x=0.0625	n=1	     log(x)=-0.0625		time=0.0003197193145751953
x=0.0625	n=2	     log(x)=-0.064453125		time=0.0005106925964355469
x=0.0625	n=3	     log(x)=-0.06453450520833333		time=0.0008575916290283203
x=0.0625	n=4	     log(x)=-0.0645383

# Taylor Summation of log<sub>e</sub>(1-x)

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">-0.03125</td>
    <td class="tg-baqh">-0.0625</td>
    <td class="tg-baqh">-0.125</td>
    <td class="tg-baqh">-0.25</td>
    <td class="tg-baqh">-0.5</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">-0.03173828125</td>
    <td class="tg-baqh">-0.064453125</td>
    <td class="tg-baqh">-0.1328125</td>
    <td class="tg-baqh">-0.28125</td>
    <td class="tg-baqh">-0.625</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">-0.031748453776041664</td>
    <td class="tg-baqh">-0.06453450520833333</td>
    <td class="tg-baqh">-0.13346354166666666</td>
    <td class="tg-baqh">-0.2864583333333333</td>
    <td class="tg-baqh">-0.6666666666666666</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">-0.031748692194620766</td>
    <td class="tg-baqh">-0.06453831990559895</td>
    <td class="tg-baqh">-0.13352457682291666</td>
    <td class="tg-baqh">-0.2874348958333333</td>
    <td class="tg-baqh">-0.6822916666666666</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">-0.03174869831457705</td>
    <td class="tg-baqh">-0.06453852113585785</td>
    <td class="tg-baqh">-0.1335313916915939</td>
    <td class="tg-baqh">-0.2876815250941685</td>
    <td class="tg-baqh">-0.6927501860119047</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">-0.0317486983145803</td>
    <td class="tg-baqh">-0.06453852113757118</td>
    <td class="tg-baqh">-0.1335313926245226</td>
    <td class="tg-baqh">-0.2876820724472973</td>
    <td class="tg-baqh">-0.6931463282650034</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">-0.0317486983145803</td>
    <td class="tg-baqh">-0.06453852113757118</td>
    <td class="tg-baqh">-0.13353139262452263</td>
    <td class="tg-baqh">-0.2876820724517809</td>
    <td class="tg-baqh">-0.6931471805530864</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">-0.0317486983145803</td>
    <td class="tg-baqh">-0.06453852113757118</td>
    <td class="tg-baqh">-0.13353139262452263</td>
    <td class="tg-baqh">-0.2876820724517809</td>
    <td class="tg-baqh">-0.6931471805599453</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">-0.0317486983145803</td>
    <td class="tg-baqh">-0.06453852113757118</td>
    <td class="tg-baqh">-0.13353139262452263</td>
    <td class="tg-baqh">-0.2876820724517809</td>
    <td class="tg-baqh">-0.6931471805599453</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">-0.0317486983145803</td>
    <td class="tg-baqh">-0.06453852113757118</td>
    <td class="tg-baqh">-0.13353139262452263</td>
    <td class="tg-baqh">-0.2876820724517809</td>
    <td class="tg-baqh">-0.6931471805599453</td>
  </tr>
  <tr>
    <td class="tg-baqh">Exact Value</td>
    <td class="tg-baqh">-0.0317486983145803</td>
    <td class="tg-baqh">-0.06453852113757118</td>
    <td class="tg-baqh">-0.13353139262452263</td>
    <td class="tg-baqh">-0.2876820724517809</td>
    <td class="tg-baqh">-0.6931471805599453</td>
  </tr>
</table>

# Taylor Summation of log<sub>e</sub>(1-x) Calculation Times

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">0.00031280517578125</td>
    <td class="tg-baqh">0.000141143798828125</td>
    <td class="tg-baqh">0.0001201629638671875</td>
    <td class="tg-baqh">0.00011110305786132812</td>
    <td class="tg-baqh">0.00010848045349121094</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.0002872943878173828</td>
    <td class="tg-baqh">0.00022983551025390625</td>
    <td class="tg-baqh">0.0002231597900390625</td>
    <td class="tg-baqh">0.00022339820861816406</td>
    <td class="tg-baqh">0.00022745132446289062</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.0004076957702636719</td>
    <td class="tg-baqh">0.0003676414489746094</td>
    <td class="tg-baqh">0.0003647804260253906</td>
    <td class="tg-baqh">0.0003581047058105469</td>
    <td class="tg-baqh">0.0003609657287597656</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">0.0006837844848632812</td>
    <td class="tg-baqh">0.0005862712860107422</td>
    <td class="tg-baqh">0.0007104873657226562</td>
    <td class="tg-baqh">0.0007562637329101562</td>
    <td class="tg-baqh">0.0006201267242431641</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">0.0010292530059814453</td>
    <td class="tg-baqh">0.0009701251983642578</td>
    <td class="tg-baqh">0.0007045269012451172</td>
    <td class="tg-baqh">0.0006501674652099609</td>
    <td class="tg-baqh">0.0006425380706787109</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0013217926025390625</td>
    <td class="tg-baqh">0.0013110637664794922</td>
    <td class="tg-baqh">0.0013127326965332031</td>
    <td class="tg-baqh">0.001291513442993164</td>
    <td class="tg-baqh">0.001359701156616211</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.002147197723388672</td>
    <td class="tg-baqh">0.001997709274291992</td>
    <td class="tg-baqh">0.001990795135498047</td>
    <td class="tg-baqh">0.0019571781158447266</td>
    <td class="tg-baqh">0.0018489360809326172</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.003262042999267578</td>
    <td class="tg-baqh">0.003217935562133789</td>
    <td class="tg-baqh">0.0030088424682617188</td>
    <td class="tg-baqh">0.0026276111602783203</td>
    <td class="tg-baqh">0.002599000930786133</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.005338191986083984</td>
    <td class="tg-baqh">0.007384538650512695</td>
    <td class="tg-baqh">0.004948616027832031</td>
    <td class="tg-baqh">0.004559993743896484</td>
    <td class="tg-baqh">0.004291534423828125</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.011217355728149414</td>
    <td class="tg-baqh">0.010558128356933594</td>
    <td class="tg-baqh">0.00966191291809082</td>
    <td class="tg-baqh">0.008798599243164062</td>
    <td class="tg-baqh">0.00824737548828125</td>
  </tr>
</table>

# Taylor Summation of log<sub>e</sub>(1-x) Absolute Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">-0.0004986983145802981</td>
    <td class="tg-baqh">-0.002038521137571178</td>
    <td class="tg-baqh">-0.008531392624522627</td>
    <td class="tg-baqh">-0.0376820724517809</td>
    <td class="tg-baqh">-0.1931471805599453</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">-1.0417064580298119e-05</td>
    <td class="tg-baqh">-8.539613757117814e-05</td>
    <td class="tg-baqh">-0.0007188926245226268</td>
    <td class="tg-baqh">-0.006432072451780901</td>
    <td class="tg-baqh">-0.06814718055994529</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">-2.4453853863376507e-07</td>
    <td class="tg-baqh">-4.015929237849436e-06</td>
    <td class="tg-baqh">-6.78509578559694e-05</td>
    <td class="tg-baqh">-0.0012237391184475865</td>
    <td class="tg-baqh">-0.026480513893278657</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">-6.119959532202568e-09</td>
    <td class="tg-baqh">-2.01231972224436e-07</td>
    <td class="tg-baqh">-6.815801605969396e-06</td>
    <td class="tg-baqh">-0.00024717661844758654</td>
    <td class="tg-baqh">-0.010855513893278657</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">-3.247402347028583e-15</td>
    <td class="tg-baqh">-1.7133239271771572e-12</td>
    <td class="tg-baqh">-9.329287342652037e-10</td>
    <td class="tg-baqh">-5.473576124015445e-07</td>
    <td class="tg-baqh">-0.00039699454804054124</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">-2.7755575615628914e-17</td>
    <td class="tg-baqh">-4.4835801737974634e-12</td>
    <td class="tg-baqh">-8.522949418443559e-07</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">-6.8588468238317546e-12</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Taylor Summation of log<sub>e</sub>(1-x) Relative Error

<style type="text/css">
.tg  {border-collapse:collapse;border-spacing:0;}
.tg td{font-family:Arial, sans-serif;font-size:14px;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg th{font-family:Arial, sans-serif;font-size:14px;font-weight:normal;padding:10px 5px;border-style:solid;border-width:1px;overflow:hidden;word-break:normal;border-color:black;}
.tg .tg-baqh{text-align:center;vertical-align:top}
</style>
<table class="tg">
  <tr>
    <th class="tg-baqh"></th>
    <th class="tg-baqh">x=1/32</th>
    <th class="tg-baqh">x=1/16</th>
    <th class="tg-baqh">x=1/8</th>
    <th class="tg-baqh">x=1/4</th>
    <th class="tg-baqh">x=1/2</th>
  </tr>
  <tr>
    <td class="tg-baqh">n=1</td>
    <td class="tg-baqh">1.595834606656954</td>
    <td class="tg-baqh">3.261633820113885</td>
    <td class="tg-baqh">6.8251140996181014</td>
    <td class="tg-baqh">15.07282898071236</td>
    <td class="tg-baqh">38.62943611198906</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=2</td>
    <td class="tg-baqh">0.032821766554539306</td>
    <td class="tg-baqh">0.1324934013225552</td>
    <td class="tg-baqh">0.5412838584640955</td>
    <td class="tg-baqh">2.286959093966543</td>
    <td class="tg-baqh">10.903548889591246</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=3</td>
    <td class="tg-baqh">0.0007702376322285691</td>
    <td class="tg-baqh">0.006222917840440589</td>
    <td class="tg-baqh">0.05083857134964341</td>
    <td class="tg-baqh">0.4271962013489757</td>
    <td class="tg-baqh">3.9720770839917985</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=4</td>
    <td class="tg-baqh">1.927625709647178e-05</td>
    <td class="tg-baqh">0.00031180230988160315</td>
    <td class="tg-baqh">0.0051045296440135265</td>
    <td class="tg-baqh">0.08599394924926228</td>
    <td class="tg-baqh">1.5910371507706123</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=8</td>
    <td class="tg-baqh">1.0228458234262711e-11</td>
    <td class="tg-baqh">2.6547306895528286e-09</td>
    <td class="tg-baqh">6.986587366811169e-07</td>
    <td class="tg-baqh">0.0001902651246799302</td>
    <td class="tg-baqh">0.057307028717812424</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=16</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">2.078580554737036e-14</td>
    <td class="tg-baqh">1.5585191442955992e-09</td>
    <td class="tg-baqh">0.0001229603197895765</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=32</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">9.895224299057006e-10</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=64</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=128</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
  <tr>
    <td class="tg-baqh">n=256</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
    <td class="tg-baqh">0.0</td>
  </tr>
</table>

# Discussion/Analysis

It can be seen that as the value of n increases, the values of the function get more and more accurate. However, this value of n is different for different values of x. Noticeably, as the value of x increases, the value of n required to get an accurate result from the function also increases. It is also noticeable that for high values of x and low values of n the function will fluctuate greatly, resulting in large relative errors.This is most visible when n is 1 and x is 4 for e<sup>x</sup> in which case the relative error is 991%. However, as the value of n increases, the relative error decreases exponentially, which follows the expected trend for the error in calculations. In terms of run time, the difference between n=1 and n=256 is about 2-3 orders of magnitude. However, the time never goes above a tenth of a second. Despite this, as the value of x increases, the value of n required to get an accurate result for the functions also increases. Due to this, as the value of x increases, the time required to calculate the value of the function accurately will also go up because it will require a larger number of iterations. Nonetheless, it can be seen from the relative error tables that for n=16, the relative errors for sin(x) and cos(x) become 0. For n=32, the relative errors for e<sup>x</sup> becomes 0 and for n=32, the relative errors for log<sub>e</sub>(1-x) becomes 0 for all values of x. Therefore, it can be deduced that the number of iterations (n) is proportional to the value of x. This means that for the tested values of x, doing 64, 128, and 256 iterations is unnecessary. However, this wouldn't be the case for larger values of x such as when x is 10<sup>10</sup>, 10<sup>20</sup> or larger. For future experiments, it is possible to try to figure out the ratio of the number of iterations required to the value of x and use that to create a function that will automatically decide on a value for n given a value for x.